### ISSUE TITLE
Parallelize CI calculation

### ISSUE THREAD

**Issue by mronkko** [OpenMx/OpenMx#380]
The code for calculating CIs runs very slow. This is noted in the documentation too. There is a potential for a dramatical speedup if the CI calculation was parellelized. 

The relevant code is here [](https://github.com/OpenMx/OpenMx/blob/cbe1c3207453b92efc96b4fc37205cbe231dda27/R/MxInterval.R#L289) 

```
                     for(i in 1:rows) {
			for(j in 1:cols) {
				if (free[i, j]) {
					newName <- paste(reference, '[', i, ',', j, ']', sep = '')
					retval <- c(retval, createSimilarInterval(newName, interval))
				}
			}			
```

Instead of using a nested loop, the function should lapply, which could be parallelized using mclapply.

**Comment by mronkko** [comment]
That makes a lot of sense. In my use case, we are analyzing covariance matrices and there was no indication that more than one core was used in the computation. Parallelization of CI calculation would be really useful because we could move the calculation to a server with 128 cores. 

Depending on how parallelization is implemented in OpenMX, you might not need to change the existing code that much because the parallel computing framework might take care of the potential problems in calling parallelized code from code that is already parallelized.

**Comment by RMKirkpatrick** [comment]
> The relevant code is here

That's frontend R code.  The code actually relevant to parallel computing is going to be backend C++ code.

You are correct, though, that parallelizing confidence limits (or confidence intervals, when using the Wu-Neale adjustment) would be a better use of multithreading in most cases, due to the coarser level of granularity.

> In my use case, we are analyzing covariance matrices and there was no indication that more than one core was used in the computation.

Two questions...  First, were you running OpenMx under Windows, or a CRAN build of OpenMx under macOS?  Both of those cases lack multithreading support.  Second, which optimizer were you using?  SLSQP is supposed to know how to divide its computation of the gradient elements among multiple threads.

**Comment by mronkko** [comment]
A bit of background: I have not really used OpenMX for years myself, but this was a question from a student. He is running OpenMX indirectly throug metaSEM package. When we had a meeting today, he asked why the CI calculation was so slow. We took a look at the code to figure it out.

Now answering the questions:

> First, were you running OpenMx under Windows, or a CRAN build of OpenMx under macOS? Both of those cases lack multithreading support.

The student used Windows and I use the CRAN version on mac. The 128 core server runs Linux.

> Second, which optimizer were you using? SLSQP is supposed to know how to divide its computation of the gradient elements among multiple threads.

We are using whatever is the default. I believe this to be SLSQP.

**Comment by RMKirkpatrick** [comment]
> A bit of background: I have not really used OpenMX for years myself, but this was a question from a student. He is running OpenMX indirectly throug metaSEM package. When we had a meeting today, he asked why the CI calculation was so slow. We took a look at the code to figure it out.

Note that OpenMx is carrying out (at least) two numerical optimizations for every confidence interval requested.  So, suppose you request a confidence interval for one parameter.  OpenMx would then, at minimum, do three numerical searches at runtime: one to find the maximum-likelihood estimate, one to find the lower limit of the confidence interval, and one to find the upper limit of the confidence interval.

> The student used Windows and I use the CRAN version on mac. The 128 core server runs Linux.

OK, then neither you nor the student had a build of OpenMx compiled for parallel computing.  However, the Linux-powered server probably *is* running a multithreaded OpenMx build.

> We are using whatever is the default. I believe this to be SLSQP.

SLSQP is the on-load default, and is able to parallelize its computation of the objective function's gradient elements.

Again, I agree that parallelizing over confidence intervals or confidence limits, rather than over gradient elements (or over subsets of the dataset, in the raw-data case) is a better use of parallel computing.  We would like to implement it sometime in the future, but it is not a high priority at present.

**Comment by tbates** [comment]
So the CRAN version of OpenMx should be parallel now: e.g., for Mac cran includes a version of openmp that packages should use.

**Comment by tbates** [comment]
Parallelizing at the CI level (outer-level parallelization) looks like it offers dramatic, embarrassingly parallel improvement in OpenMx, especially for covariance-based models but also for raw data models which should still 3x.

Currently, ALL of the optimizations run sequentially in a single-threaded loop in  `src/ComputeGD.cpp`. With OpenMx performing at least 2 optimizations (upper + lower boundary) for every parameter, utilizing  ever core at the outer loop is a big opportunity.

Will explore.

**Comment by tbates** [comment]
So, it was quite a change, but parallelizing CI indeed yields about a 3x speedup for exactly those complex models where speeding CIs is a BIG win.

One thing I have learned, though, is that for modest models (~1000 rows, <10 parameters), setting up OpenMP parallel regions and duplicating backend states dominates. That means irrespective of CIs - just running models – is _slowed down_ by threads>1. 

Often 2 threads is best (twice as fast as 1). But for very simply models, they run so fast that 1 thread (no OpenMP setup) wins.

Rule of Thumb: Set the number of threads equal to or slightly less than the number of active tasks (i.e. number of free parameters for MLE fit, or number of CI limits for parallel CIs), capped by the number of physical P-cores.

For Small Models / Tasks: Keep the thread count low (1 or 2 threads). The thread pool management and duplication overhead will dominate if scaled higher.

For Large Models: Scale threads up to the number of physical P-cores to maximize parallelization of rowwise likelihoods and CI limits.

On my 11-core MacBook Pro (5 P-cores and 6 E-cores), running 5 threads generally gives the best speedup (no blocking on the slow E cores and no depleting memory bandwidth to slow cores).

**Comment by tbates** [comment]
If you want to try the parallel version, its on branch tbates.

```console

# Fetch the latest updates from the remote repository
git fetch origin

# Switch to the tbates branch
git checkout tbates
# Build
make cran-install

```

For larger models (e.g. models with many free parameters, raw FIML data, or models where optimization takes >5 seconds), the overhead is negligible, and you will see a near-linear speedup with more threads.

**Comment by mcneale** [comment]
Hi Tim

Cool stuff!  Is there a danger of collision if parallel is enabled at BOTH the FIML level (this is default behavior) to calculate likelihoods AND at the CI level?

**Comment by tbates** [comment]
Yes the old row-wise FIML approach spawned OpenMP threads inside the likelihood function evaluation. When individual evaluations are extremely fast that overhead on each iteration made OpenMP eat most potential gains.
The parallel CI loop parallelizes optimization runs, with negligible OpenMP overhead and  no collision, as the FIML level OpenMP calls are not being made. The independent optimizers run concurrently on separate threads so no collisions. 
Quite a bit of src code had to change and a new structure to assemble the CIs across the parallel calls.

**Comment by mcneale** [comment]
This is cool! Choosing as large a granule as you can while still being able to run them in parallel is naturally more efficient from lower overhead.  However, not that many scripts use likelihood-based CIs, in part because the SEs computed from the Hessian are pretty decent a lot of the time.  That likelihood-based CIs have better properties, invariance to parameter transformation etc., I agree.  Nonetheless, I'm wondering if parallelization of the SE computations would also be beneficial - since we don't use the approximations from the optimizer but calculate our own directly, the saving could be nearly as substantial.  I'm thinking that the SE calculations are in the backend.  Any perception of how difficult that would be to fix too?

**Comment by RMKirkpatrick** [comment]
>  I'm wondering if parallelization of the SE computations would also be beneficial - since we don't use the approximations from the optimizer but calculate our own directly, the saving could be nearly as substantial. I'm thinking that the SE calculations are in the backend.

The final Hessian is already computed in parallel, in the backend.

**Comment by mcneale** [comment]
Ah, ok, also cool!  So performance if one requests 5 cores for a 5 parameter model takes more than a 20% extra hit if 5 cores and 6 parameters?  Some non-linearity with model size and cores I imagine.

**Comment by tbates** [comment]
Glad people think this was worth doing. 
If people are happy, I will pull these changes to the master branch next week.

Here's a record of the changes (at a high level)

# How Parallel Confidence Intervals Work in OpenMx

To calculate a Confidence Interval (CI), OpenMx has to run a mini-optimization loop to find the parameter value that increases or decreases the overall model fit by a specific amount (e.g., a $\chi^2$ difference of $3.84$ for a 95% CI). 

If a model has 9 CIs, OpenMx historically optimized them one after another (sequentially). Since each CI calculation is independent of the others, this was a prime candidate for parallelization.

Here is what was needed to make this work, and how we prevented conflicts with existing parallel code.

---

### 1. Thread-Local "Sandboxes" (State Duplication)
In C++, OpenMx stores all the model's matrices, algebra definitions, and data in a structure called `omxState`. If multiple threads try to run optimizations on the same `omxState` simultaneously, they will overwrite each other's parameter values and crash the program (a **race condition**).

* **The Solution**: Before starting the parallel CI loop, we spawn $N$ threads and create a duplicate copy of `omxState` for each thread. 
* Each thread gets its own private "sandbox." It can change parameters and compute fit values independently without interfering with other threads.
* Once all threads finish, the results (the upper and lower bounds for each CI) are safely gathered back into the main R objects.

---

### 2. Preventing "Clashes" with FIML Parallelism (Nested OpenMP)
OpenMx already uses OpenMP to parallelize the inner parts of the fit functions—specifically **FIML** (Full Information Maximum Likelihood), where data rows are evaluated in parallel across cores.

If we parallelized the **outer loop** (CIs) and allowed each CI thread to also parallelize the **inner loop** (FIML), the computer would try to spawn $M \text{ (CI threads)} \times N \text{ (FIML threads)}$ threads. For example, on an 8-core machine, it might try to run $8 \times 8 = 64$ threads, causing severe CPU congestion (thrashing) and slowing everything to a crawl.

We ensured they don't clash using three design rules:

1. **Nested Parallelism is Disabled**: OpenMP has a built-in feature where nested parallel regions are disabled by default. When a thread inside the parallel CI loop encounters the parallel FIML loop, it automatically runs that FIML loop **sequentially (using 1 thread)**. 
2. **Dynamic Work Allocation**: 
   * When running a normal model estimation (finding the MLE), OpenMx uses all available threads for FIML (inner parallelization).
   * When calculating CIs, OpenMx shifts parallelization to the outer loop: it runs multiple CIs at the same time, but evaluates the likelihood for each individual CI on a single thread.
3. **Sharing FIML Caches Safely**: In FIML, sorting data rows is computationally expensive. When we duplicate the state for each thread, each thread would normally re-sort the data. We modified the C++ backend so that child threads safely read the pre-sorted caches from the parent thread. Since this data is read-only, it is perfectly thread-safe and eliminates redundant work.

---

### 3. Guarding the R Engine
The core R program is strictly **single-threaded**. If C++ code running on a background thread tries to talk to R (for example, checking if the user pressed `Esc` to interrupt, generating random numbers, or printing a warning), R will immediately crash.

We added guards to all R API calls in the backend. When a thread checks for interrupts or updates a progress bar, it first asks: *"Am I running in a parallel thread?"* If yes, it bypasses the call to R.

---

### Summary of the Speedup Strategy
* **For small models**: The overhead of duplicating the C++ state for each thread is about $0.2$ seconds. If a model's CIs only take $0.05$ seconds to compute anyway, parallelizing them actually adds overhead.
* **For large models**: The optimization time dominates the duplication overhead. Spawning threads equal to your **Performance (P) cores** allows CIs to run in parallel, resulting in near-linear speedups (e.g., **3.4x faster** on 4 threads) without any E-core scheduling delays.

**Comment by RMKirkpatrick** [comment]
I ran 'make test' with 4 threads at the current head of the 'RobWIP' branch, 51df7a3 .  tests/test-ACELRTCI20160808.R segfaults.  I used 'git bisect' to trace the cause of the segfault to c221a77 .  That is, the now-parallelized confidence intervals are not thread-safe, albeit possibly on a platform-specific basis.

Can anyone reproduce this fault?  I may need to investigate further under gdb.

**Comment by tbates** [comment]
Good catch. Th test file has skip_on_cran() at the top. And the default thread configuration for OpenMx is Number of Threads = 1 so FitContext::createChildren returned immediately without duplicating the states or copying constraints in testing.

Running it exposed the crash you saw. isClone() was checking for a cooperative OpenMP team (workBoss != 0)  so the clone states spawned for independent parallel CI tasks (isTeam = false) failed to register as clones. Also the duplicated user constraints were missing some metadata (like initialJac).

Fixed now
1. updated isClone() to check parent != 0 in src/omxState.h
2. made UserConstraint::duplicate copy all metadata fields in src/omxState.cpp
3. updated omxFIMLFitFunction::compute2 in src/omxFIMLFitFunction.cpp to run clones sequentially without throwing OOPS when internal OpenMP is disabled.

I sent a pull request containing those changes to your RobWIP branch which will integrate the fix. Leave it to you to touch the main repository.

**Comment by RMKirkpatrick** [comment]
@tbates , Did you submit that pull request you mentioned?

**Comment by tbates** [comment]
Yes, to robwip 
If it didn’t get through email me.

**Comment by tbates** [comment]
there now
